chain/consensus/hierarchical/actors/sca/sca_apply.go

```diff
@@ -41,6 +52,67 @@ func applyFund(rt runtime.Runtime, msg types.Message) {
 	}
 }
 
+func applyRelease(rt runtime.Runtime, msg types.Message) {
+	var st SCAState
+
+	rfrom, rto := fromToRawAddr(rt, msg.From, msg.To)
+	snFrom, err := msg.From.Subnet()
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "error getting subnet from HAddress")
+
+	// Check that msg.From == BurntRewardsActorAddr, this determines that is a release message
+	if rfrom != builtin.BurntFundsActorAddr {
+		rt.Abortf(exitcode.ErrIllegalArgument, "msg passed as argument not a funding message (from != burtnRewardActorAddr)")
+	}
+
+	rt.StateTransaction(&st, func() {
+		bottomUpStateTransition(rt, &st, msg)
+
+		// Update circulating supply reducing release value.
+		sh, has, err := st.GetSubnet(adt.AsStore(rt), snFrom)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "error fetching subnet state")
+		if !has {
+			rt.Abortf(exitcode.ErrIllegalState, "subnet for actor hasn't been registered yet")
+		}
+		// Even if the actor has balance, we shouldn't allow releasing more than
+		// the current circulating supply, as it would mean that we are
+		// releasing funds from the collateral.
+		if sh.CircSupply.LessThan(msg.Value) {
+			rt.Abortf(exitcode.ErrIllegalState, "wtf! we can't release funds below the circulating supply. Something went wrong!")
+		}
+		sh.CircSupply = big.Sub(sh.CircSupply, msg.Value)
+		st.flushSubnet(rt, sh)
+	})
+
+	// Release funds to the destination address.
+	code := rt.Send(rto, builtin.MethodSend, nil, msg.Value, &builtin.Discard{})
+	if !code.IsSuccess() {
+		noop(rt, code)
+	}
+}
+
+func bottomUpStateTransition(rt runtime.Runtime, st *SCAState, msg types.Message) {
+	// Bottom-up messages include the nonce of their message meta. Several messages
+	// will include the same nonce. They need to be applied in order of nonce.
```

Comment:
the same nonce? this can cause arbitrary re-ordering of these messages breaking determinism

Reply:
We use the same nonce for cross-msgs include in the same `CrossMsgMeta` to identify that they belong to the "same package". This ensures the ordering between cross-msg propagated by several child subnets. However, before being applied, the original nonce of the message is recovered (through the cross-msg resolution protocol) and all messages are sorted conveniently. Thus, the order of application is deterministic, and someone trying to forge it would reach an inconsistent state.

That being said, the comment is really on point, the code that handles message sorting is outside the actor (it is [here](https://github.com/filecoin-project/eudico/pull/94/files#diff-96caa44c6fa84d9d0cf1e6b9820105e89227da2112d9ed8653727dfc6af5718dR272), in case you want to have a quick look at it)